quixstreams/sources/community/pubsub/pubsub.py

```diff
@@ -0,0 +1,126 @@
+        subscription_id: str,
+        service_account_json: Optional[str] = None,
+        commit_every: int = 100,
+        commit_interval: float = 5.0,
+        create_subscription: bool = False,
+        shutdown_timeout: float = 10.0,
+    ):
+        """
+        :param project_id: a Google Cloud project ID.
+        :param topic_id: a Pub/Sub topic ID (NOT the full path).
+        :param subscription_id: a Pub/Sub subscription ID (NOT the full path).
+        :param service_account_json: a Google Cloud Credentials JSON as a string
+            Can instead use environment variables:
+            - "GOOGLE_APPLICATION_CREDENTIALS" set to a JSON filepath
+            - "PUBSUB_EMULATOR_HOST" set to a URL if using an emulated Pub/Sub
+        :param commit_every: max records allowed to be processed before committing.
+        :param commit_interval: max allowed elapsed time between commits.
+        :param create_subscription: whether to attempt to create a subscription at
+            startup; if it already exists, it instead logs its details (DEBUG level).
+        :param shutdown_timeout: How long to wait for a graceful shutdown of the source.
+        """
+        self._credentials = service_account_json
+        self._project_id = project_id
+        self._topic_id = topic_id
+        self._subscription_id = subscription_id
+        self._commit_every = commit_every
+        self._commit_interval = commit_interval
+        self._create_subscription = create_subscription
+        super().__init__(
+            name=subscription_id,
+            shutdown_timeout=shutdown_timeout,
+        )
+
+    def default_topic(self) -> Topic:
+        return Topic(
+            name=f"pubsub_{self._subscription_id}_{self._topic_id}",
+            key_deserializer="str",
+            value_deserializer="bytes",
+            key_serializer="str",
+            value_serializer="bytes",
+        )
+
+    def _handle_pubsub_message(self, message: PubSubMessage):
+        timestamp: datetime.datetime = message.publish_time
+        kafka_msg = self.serialize(
+            key=message.ordering_key,  # an empty string if ordering not enabled
+            value=message.data,
+            timestamp_ms=int(timestamp.timestamp() * 1000),
+        )
+        self.produce(
+            key=kafka_msg.key,
+            value=kafka_msg.value,
+            timestamp=kafka_msg.timestamp,
+        )
+
+    def run(self):
+        with PubSubConsumer(
+            credentials=self._credentials,
+            project_id=self._project_id,
+            topic_id=self._topic_id,
+            subscription_id=self._subscription_id,
+            create_subscription=self._create_subscription,
+            async_function=self._handle_pubsub_message,
+            default_poll_timeout=self._commit_interval,
+            max_batch_size=self._commit_every,
+        ) as consumer:
+            while self._running:
+                consumer.poll()
+                self.flush()
```

Comment:
Does it flush on every received message?

Reply:
The polling is actually handing a batch of messages at a time. 

I tweaked the name of the function to better reflect that behavior (which is polling + producing the message via a callback on the async consumer).